### ISSUE TITLE
No character metrics for emoji characters

### ISSUE THREAD

**Issue by jameshfisher** [KaTeX/KaTeX#3718]
**Describe the bug:**
A clear and concise description of what the bug is. If your installation is available on the web or you can provide a minimal repo or CodePen that reproduces the bug, please add the link to it.

**To Reproduce:**
1. Go to [katex.org](https://katex.org/)
2. Paste the LaTeX string: `\sqrt{\char"1F642} = \text{?}`
3. Open dev console
4. See warning: `No character metrics for '🙂' in style 'Main-Regular' and mode 'math'`
5. See rendering where "🙂" character is rather larger than the space allocated for it:

<img width="598" alt="image" src="https://user-images.githubusercontent.com/166966/191703165-fc47afd4-6921-4e31-84f4-efe14ebeeb6b.png">

**Expected behavior:**
The emoji character "🙂" should be rendered under a square root symbol. The size of the character should be accounted for. No logs should be seen in the dev console.

**Environment (please complete the following information):**
 - KaTeX Version: 0.16.2
 - Device: Desktop
 - OS: Microsoft Windows 11 Home
 - Browser: Chrome 
 - Version: 105.0.5195.127 (Official Build) (64-bit)

**Additional context:**
The same problem seems to happen with any other emoji character.

**Comment by jameshfisher** [comment]
Note this also happens for the non-emoji character "☺", which is used in [the official example for the `\char` function](https://katex.org/docs/supported.html). However, it renders more sensibly (due to its smaller size).

**Comment by jameshfisher** [comment]
This is the core logic for finding character metrics: https://cs.github.com/KaTeX/KaTeX/blob/ef49e2be91628e96e4fcb962b28b443930627d84/src/fontMetrics.js#L211-L249

Here we have this interesting feature:

```
        // We don't typically have font metrics for Asian scripts.
        // But since we support them in text mode, we need to return
        // some sort of metrics.
        // So if the character is in a script we support but we
        // don't have metrics for it, just use the metrics for
        // the Latin capital letter M. This is close enough because
        // we (currently) only care about the height of the glpyh
        // not its width.
```

Perhaps we can do a similar thing for emoji characters? They are all similar height, and we can check whether it's in an emoji character range.

**Comment by jameshfisher** [comment]
Or more generally, can KaTeX use [the TextMetrics API](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics) for fallback characters?:

```
> const canvas = new OffscreenCanvas(1, 1);
> const ctx = canvas.getContext('2d');
> const metrics = ctx.measureText('🙂');
TextMetrics {
  width: 13.73046875, 
  actualBoundingBoxLeft: -1, 
  actualBoundingBoxRight: 12, 
  actualBoundingBoxAscent: 9, 
  actualBoundingBoxDescent: 2, 
  …
}
```

I'm not sure how/whether these numbers correspond to KaTeX's metrics, though ...

**Comment by jameshfisher** [comment]
Hmm, I haven't actually made my primary pain clear: my server-side logs are littered with thousands of lines like

```
[2022-09-22 07:24:52] No character metrics for '🐦' in style 'Main-Regular' and mode 'text'
[2022-09-22 07:24:52] No character metrics for '🦅' in style 'Main-Regular' and mode 'text'
```

That's due to [this TODO](https://github.com/KaTeX/KaTeX/blob/main/src/buildCommon.js#L74-L76)

```
// TODO(emily): Figure out a good way to only print this in development
```

I'd be reasonably happy if I could just get these lines to go away in production.

**Comment by jameshfisher** [comment]
(I just opened https://github.com/KaTeX/KaTeX/issues/3720 for the logging issue, as I think it's a different problem.)

**Comment by edemaine** [comment]
KaTeX only has and uses metrics for its own fonts, hence the error. For server-side rendering (which you're using it seems), we can't use the TextMetrics API (even if present, we wouldn't know what fonts the client would use).

> Note this also happens for the non-emoji character "☺" (`\char"263a`), which is used in [the official example for the `\char` function](https://katex.org/docs/supported.html).

Perhaps we should add a note in the documentation there about character metrics so the vertical spacing is likely off.

> Perhaps we can do a similar thing for emoji characters? They are all similar height, and we can check whether it's in an emoji character range.

Is this actually true (reference?), or does the relative height depend on the specific emoji font / operating system? Even if it's approximately true, this would seem like a good fix for emoji.

**Comment by jameshfisher** [comment]
> > Perhaps we can do a similar thing for emoji characters? They are all similar height, and we can check whether it's in an emoji character range.
> 
> Is this actually true (reference?), or does the relative height depend on the specific emoji font / operating system? Even if it's approximately true, this would seem like a good fix for emoji.

It does seem at least approximately true that, for a given emoji font, all characters are the same width and height. E.g. see the grids of emojis at https://getemoji.com/. It's not formatted as a grid, they're just all the same width. Hacky way to check the heights are also the same: set `writing-mode: vertical-rl`, and observe that they're still all in a grid.

The width+height certainly seems to vary by font, but perhaps we can just pick a size that approximates common emoji fonts?

**Comment by edemaine** [comment]
> The width+height certainly seems to vary by font, but perhaps we can just pick a size that approximates common emoji fonts?

Yes, this seems reasonable. I think it would be useful to do something like this, given that KaTeX isn't ever going to include its own emoji font, but supporting them at least a little better would seem helpful.

I'm not sure how to easily sample emoji fonts, as several of them are operating system specific. If you could make a TextMetrics demo page (you pretty much already have), I could run it on various operating systems via BrowserStack. We can of course also look directly at free fonts like [Noto Emoji](https://fonts.google.com/noto#emoji-qaae-color), [OpenMoji](https://openmoji.org/), and [Twemoji](https://twemoji.twitter.com/)
